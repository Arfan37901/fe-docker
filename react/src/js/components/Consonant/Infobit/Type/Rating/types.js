import {
    string,
    number,
} from 'prop-types';

// eslint-disable-next-line import/prefer-default-export
export const RatingType = {
    label: string,
    totalStars: number,
    starsFilled: number,
};
