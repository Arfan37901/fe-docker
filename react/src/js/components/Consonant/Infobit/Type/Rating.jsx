import React from 'react';
// import PropTypes from 'prop-types';

const Rating = () => (
    <span>Rating bit</span>
);

export default Rating;
