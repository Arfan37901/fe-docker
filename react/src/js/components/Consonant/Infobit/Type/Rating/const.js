export const maxAllowedStars = 5;
export const maxPercantage = 100;
export const defaultFilledStars = 0;
