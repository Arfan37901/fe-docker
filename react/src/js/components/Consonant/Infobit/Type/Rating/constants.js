export const scalingFactor = 100;
export const maxAllowedStars = 5;

export const defaultFilledStars = 0;
export const defaultTotalStars = Number.MAX_SAFE_INTEGER;


export const defaultProps = {
    label: '',
    totalStars: maxAllowedStars,
    starsFilled: defaultFilledStars,
};
