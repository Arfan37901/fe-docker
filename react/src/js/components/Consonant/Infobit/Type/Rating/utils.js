import {
    scalingFactor,
    maxAllowedStars,
    defaultTotalStars,
    defaultFilledStars,
} from './constants';

/**
* Will calculate the rating (percentage) that is shown
* @param {Number} starsFilled - How many stars should be filled
* @param {Number} totalStars - Total amount of stars to display
* @returns {Number} - Rating Percentage
*/
// eslint-disable-next-line import/prefer-default-export
export const getRating = (starsFilled, totalStars) => {
    const starsFilledNotValid = starsFilled < 0 || starsFilled > totalStars;
    const totalStarsNotValid = totalStars <= 0 || totalStars > maxAllowedStars;

    const filled = starsFilledNotValid
        ? defaultFilledStars
        : starsFilled;

    const total = totalStarsNotValid
        ? defaultTotalStars
        : totalStars;

    const ratingPercentage = filled / total;
    const rating = Math.round(ratingPercentage * scalingFactor);

    return rating;
};
