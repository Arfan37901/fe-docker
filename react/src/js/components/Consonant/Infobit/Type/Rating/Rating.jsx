import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { maxAllowedStars, maxPercantage, defaultFilledStars } from './const';

function Rating({ totalStars, starsFilled, label }) {
    let total = totalStars;
    let filled = starsFilled;
    const className = classNames('consonant-rating-infobit', { 'consonant-rating-infobit_neg-margin': !label });

    if (totalStars <= 0 || totalStars > maxAllowedStars) {
        total = maxAllowedStars;
        filled = totalStars > 0 && starsFilled >= 0 ?
            Math.round((starsFilled * maxAllowedStars) / totalStars) :
            defaultFilledStars;
    }

    if (filled < 0 || filled > total) filled = defaultFilledStars;

    const starsFilledMultiplier = Math.round(maxPercantage / total);

    return (
        <div
            className={className}
            data-stars={total}>
            <span
                data-testid="rating-star"
                className="consonant-rating-infobit--stars"
                data-rating={Math.round(filled * starsFilledMultiplier)} />
            {label && <span className="consonant-rating-infobit--text">{label}</span>}
        </div>
    );
}

Rating.propTypes = {
    totalStars: PropTypes.number,
    starsFilled: PropTypes.number,
    label: PropTypes.string,
};

Rating.defaultProps = {
    totalStars: maxAllowedStars,
    starsFilled: defaultFilledStars,
    label: '',
};

export default Rating;
