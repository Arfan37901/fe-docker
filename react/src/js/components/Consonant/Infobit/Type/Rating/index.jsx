import React, { useMemo } from 'react';
import classNames from 'classnames';

import { getRating } from './utils';
import { RatingType } from './types';
import { If } from '../../../Common';
import { defaultProps } from './constants';
import { parseToPrimitive } from '../../../Helpers/general';

/**
 * Rating Infobit (shown in 3:2 Card Footer)
 *
 * Displays how many stars should be shown for a giving rating
 *
 * @component
 * @example
 * const props= {
    totalStars: Number,
    starsFilled: Number,
    label: String,
 * }
 * return (
 *   <Rating {...props} />
 * )
 */
const Rating = ({
    label,
    totalStars,
    starsFilled,
}) => {
    const className = classNames({
        'consonant-rating-infobit': true,
        'consonant-rating-infobit_neg-margin': !label,
    });

    const ratingToDisplay = useMemo(() => {
        const total = parseToPrimitive(totalStars);
        const filled = parseToPrimitive(starsFilled);

        return getRating(filled, total);
    }, [starsFilled, totalStars]);

    return (
        <div
            className={className}
            data-stars={totalStars}>
            <span
                data-testid="rating-star"
                data-rating={ratingToDisplay}
                className="consonant-rating-infobit--stars" />
            <If condition={Boolean(label)}>
                <span
                    className="consonant-rating-infobit--text">
                    {label}
                </span>
            </If>
        </div>
    );
};

Rating.propTypes = RatingType;
Rating.defaultProps = defaultProps;

export default Rating;
