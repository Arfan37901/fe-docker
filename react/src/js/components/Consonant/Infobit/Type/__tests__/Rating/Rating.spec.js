import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';

import Rating from '../../Rating/Rating';

describe('Consonant/Rating', () => {
    test('should correct render', async () => {
        const tree = renderer
            .create(<Rating />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
    test('should correct render with totalStars', async () => {
        const tree = renderer
            .create(<Rating totalStars={10} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
    test('should correct render with totalStars < 0', async () => {
        const tree = renderer
            .create(<Rating totalStars={-1} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
