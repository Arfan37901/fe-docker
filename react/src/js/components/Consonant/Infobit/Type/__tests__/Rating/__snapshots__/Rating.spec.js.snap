// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consonant/Rating should correct render 1`] = `
<div
  className="consonant-rating-infobit consonant-rating-infobit_neg-margin"
  data-stars={5}
>
  <span
    className="consonant-rating-infobit--stars"
    data-rating={0}
  />
  
</div>
`;

exports[`Consonant/Rating should correct render with totalStars < 0 1`] = `
<div
  className="consonant-rating-infobit consonant-rating-infobit_neg-margin"
  data-stars={5}
>
  <span
    className="consonant-rating-infobit--stars"
    data-rating={0}
  />
  
</div>
`;

exports[`Consonant/Rating should correct render with totalStars 1`] = `
<div
  className="consonant-rating-infobit consonant-rating-infobit_neg-margin"
  data-stars={5}
>
  <span
    className="consonant-rating-infobit--stars"
    data-rating={0}
  />
  
</div>
`;
